apidocs: Move stray descriptions to OpenAPI.

MSurfer20 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Testing plan:** <!-- How have you tested? -->


**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

MSurfer20 commented:
@timabbott This should be ready for review after the prep PR is merged. This step crosses the last hurdle in creating a common template, and I think we have reached maximum coverage once this gets merged :)

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -9507,6 +9507,18 @@ paths:
       summary: Get server settings
       tags: ["server_and_organizations"]
+      x-response-description: |
+        [ldap-auth]: https://zulip.readthedocs.io/en/latest/production/authentication-methods.html#ldap-including-active-directory
```
Does this link go somewhere else in the page?  That seems likely.

MSurfer20 replied:
@timabbott That's indeed the case. Moving it to correct location.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -10264,6 +10264,8 @@ paths:
               You may pass in one or more of the parameters mentioned above
               as URL query parameters, like so:
+      x-parameter-description: |
+        **Note**: The following parameters are all URL query parameters.
```
I think we should just delete this one; all HTTP GET requests are supposed to be encoded this way, and we cover this in the examples I think.

MSurfer20 replied:
It indeed feels little redundant -- removed this description.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -3867,6 +3867,13 @@ paths:
                 allOf:
                   - $ref: "#/components/schemas/BadEventQueueIdError"
                   - description: |
+                      #### BAD_EVENT_QUEUE_ID errors
```
Can you confirm the heading works properly?  I'm worried about leading whitespace confusing markdown.

MSurfer20 replied:
I had checked it earlier, and it was indeed working properly.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -3867,6 +3867,13 @@ paths:
+                      handle these correctly.
+
                       If the target event queue has been garbage collected, you'll get the
                       following error response
```
This last sentence seems awkward, but maybe that's not new.

MSurfer20 replied:
Modified the descriptions -- might be little better now :)

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -7033,6 +7033,8 @@ paths:
             description: |
               You may pass the `client_gravatar` or `include_custom_profile_fields` query parameter as follows:
+      x-parameter-description: |
+        **Note**: The following parameters are all URL query parameters.
```
Yeah, I think we should just delete these "URL query parameter" reminders.  If we want to have a notice of this form, we should compute whether to do it computationally.

timabbott replied:
Can you start a thread in "#documentation" about these notices?  It's probably worth designing what we want to replace them with and opening an issue for our proposed solution.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -5019,6 +5019,14 @@ paths:
 
         `{msg_id}` in the above URL should be replaced with the ID of the
         message you wish you update.
+        ## Permissions
```
I think we can skip this heading, unless there's a link to it.  Also, this information is probably inaccurate; we should instead just link to the /help/ article on message editing permissions: https://zulip.com/help/configure-message-editing-and-deletion.  So maybe we just have a sentence:

`See [configuring message editing](/help/configure-message-editing-and-deletion) for detailed documentation on under what circumstances users are allowed to edit topics."

MSurfer20 replied:
Made the appropriate modifications.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -4224,8 +4224,12 @@ paths:
+                    description: |
+                      When a request is successful, this endpoint returns a dictionary
+                      containing the following (in addition to the `msg` and `result` keys
+                      present in all Zulip API responses).
```
Question: Do we need this paragraph?  I would sorta expect that the `msg` and `result` keys would appear in the documentation these days, and it is just legacy content that we can delete.  (It certainly doesn't seem to add any extra detail).

MSurfer20 replied:
They do clearly appear in examples, but not in parameters from what I can see, and they are being mentioned here instead.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -6571,6 +6571,25 @@ paths:
+        The emoji name can only contain letters, numbers, dashes, and spaces.
+        Upper and lower case letters are treated the same, and underscores (_)
+        are treated the same as spaces (consistent with how the Zulip UI
+        handles emoji).
```
I think this paragraph belongs in the `parameters` description for the `emoji_name` field.

MSurfer20 replied:
Hmm, weirdly enough, the parameters table wasn't rendered for this endpoint in its `.md` file. Moving the description to appropriate place and fixing template.

MSurfer20 commented:
@timabbott Addressed most of the comments above. I think the PR would be easier to review after the prep one gets merged(since the things can't be verified directly on this PR as of now).

timabbott commented:
I just merged the prep commit, so you can rebase this now.  It'd be great if you can post the HTML diff output after completing this one.

(As a sidenote, did you end up writing a tool for that sort of HTML diff work?  I wonder if it'd be worth checking it and documenting it).

MSurfer20 commented:
@timabbott Done with the rebase. Attaching the diff of the files:
```
old-archive-stream new-archive-stream
246d245
< <p><strong>Note</strong>: The following parameters are all URL query parameters.</p>

old-get-events new-get-events
320,324d319
< <p><strong>Note</strong>: The parameters documented above are optional in the sense that
< even if you haven't registered a queue by explicitly requesting the
< <code>http://localhost:9991/api/v1/register</code> endpoint, you could pass the parameters for
< <a href="/api/register-queue">the <code>http://localhost:9991/api/v1/register</code> endpoint</a> to this
< endpoint and a queue would be registered in the absence of a <code>queue_id</code>.</p>
4534,4535c4529,4534
< <p>If the target event queue has been garbage collected, you'll get the
< following error response:</p>
---
> <p>This error occurs if the target event queue has been garbage collected.
> A compliant client will handle this error by re-initializing itself
> (e.g. a Zulip web app browser window will reload in this case).</p>
> <p>See <a href="/api/register-queue">the /register endpoint docs</a> for details on how to
> handle these correctly.</p>
> <p>The following is the error response in such case:</p>
4543,4546d4541
< <p>A compliant client will handle this error by re-initializing itself
< (e.g. a Zulip web app browser window will reload in this case).</p>
< <p>See <a href="/api/register-queue">the /register endpoint docs</a> for details on how to
< handle these correctly.</p>

old-get-messages new-get-messages
402,404d401
< <p>When a request is successful, this endpoint returns a dictionary
< containing the following (in addition to the <code>msg</code> and <code>result</code> keys
< present in all Zulip API responses).</p>
601c598,601
< <p>A typical successful JSON response may look like:</p>
---
> <p>When a request is successful, this endpoint returns a dictionary
> containing the following (in addition to the <code>msg</code> and <code>result</code> keys
> present in all Zulip API responses).
> A typical successful JSON response may look like:</p>

old-get-stream-id new-get-stream-id
263d262
< <p><strong>Note</strong>: The following parameters are all URL query parameters.</p>

old-get-streams new-get-streams
270d269
< <p><strong>Note</strong>: The following parameters are all URL query parameters.</p>

old-get-user new-get-user
257d256
< <p><strong>Note</strong>: The following parameters are all URL query parameters.</p>

old-get-user-by-email new-get-user-by-email
264d263
< <p><strong>Note</strong>: The following parameters are all URL query parameters.</p>

old-get-users new-get-users
276d275
< <p><strong>Note</strong>: The following parameters are all URL query parameters.</p>

old-update-message new-update-message
218a219,220
> <p><strong>Note</strong>: See <a href="/help/configure-message-editing-and-deletion">configuring message editing</a> for detailed documentation on under what
> circumstances users are allowed to edit topics.</p>
278,285d279
< <h2 id="permissions">Permissions</h2>
< <p>You only have permission to edit a message if:</p>
< <ol>
< <li>You sent it, <strong>OR</strong>:</li>
< <li>This is a topic-only edit for a (no topic) message, <strong>OR</strong>:</li>
< <li>This is a topic-only edit and you are an admin, <strong>OR</strong>:</li>
< <li>This is a topic-only edit and your realm allows users to edit topics.</li>
< </ol>

old-upload-custom-emoji new-upload-custom-emoji
246c246
< <div class="codehilite" data-code-language="curl"><pre><span></span><code>curl -sSX POST http://localhost:9991/api/v1/realm/emoji/demo \
---
> <div class="codehilite" data-code-language="curl"><pre><span></span><code>curl -sSX POST http://localhost:9991/api/v1/realm/emoji/smile \
254,257c254,260
< <p>As described above, the image file to upload must be provided in the
< request's body.</p>
< <h2 id="emoji-name">Emoji name</h2>
< <p>The emoji name can only contain letters, numbers, dashes, and spaces.
---
> <div class="api-argument" id="parameter-emoji_name">
>     <p class="api-argument-name"><strong>emoji_name</strong> <span class="api-field-type">string</span> <span class="api-argument-required">required in path</span>  <a href="#parameter-emoji_name" class="api-argument-hover-link"><i class="fa fa-chain"></i></a></p>
>     <div class="api-example">
>         <span class="api-argument-example-label">Example</span>: <code>&quot;smile&quot;</code>
>     </div>
>     <div class="api-description"><p>The name that should be associated with the uploaded emoji image/gif.
> The emoji name can only contain letters, numbers, dashes, and spaces.
260c263,268
< handles emoji).</p>
---
> handles emoji).</p></div>
>     <hr>
> </div>
> 
> <p>As described above, the image file to upload must be provided in the
> request's body.</p>
```
Also, I'm using combination of a python script and a bash script to store raw HTMLs in files `old-<endpoint-name>` and `new-<endpoint-name>`, and showing diff of the respective older and newer files if they not exactly same(using cmp bash command).

timabbott commented:
For `upload-custom-emoji`, I think `smile` is a terrible example, since that's a built-in emoji name, and custom emoji are primarily for adding new custom emoji. We can merge this, but we should change that example to something that's a reasonable custom emoji name; I think `demo` (as was previously present` is OK, but `my_new_emoji` might be better.

timabbott commented:
>  The emoji name can only contain letters, numbers, dashes, and spaces.

Is this accurate?  I would have thought that we used underscores, not `-`s, for this purpose.

timabbott commented:
Merged after a small whitespace change for the "delete message" one to add a blank line after the heading.  Thanks @MSurfer20!

Can you take care of the small follow-up issues mentioned in my last batch of comments?  

It's exciting how close we are to being able to delete the templates.
